### PR TITLE
Add base sensor

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
- Add a BaseSensor clas.  People can inherit BaseSensor to implement their sensors. 
- Add a file sensor into `flytekit.core`, which uses fsspec to check if file exists

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [x] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
```python
import asyncio
from typing import Optional

import fsspec
import grpc
from fsspec.utils import get_protocol

from flytekit.configuration import DataConfig
from flytekit.core.data_persistence import s3_setup_args
from flytekit.core.sensor_task import FileSensorConfig
from flytekit.extend.backend.base_agent import AgentRegistry
from flytekit.extend.backend.base_sensor import SensorBase
from flytekit.models.literals import LiteralMap
from flytekit.models.task import TaskTemplate


class FileSensor(SensorBase):
    def __init__(self):
        super().__init__(task_type="file_sensor")

    async def poke(self, cfg: FileSensorConfig) -> bool:
        path = cfg.path
        protocol = get_protocol(path)
        kwargs = {}
        if get_protocol(path):
            kwargs = s3_setup_args(DataConfig.auto().s3, anonymous=False)
        fs = fsspec.filesystem(protocol, **kwargs)
        return await asyncio.to_thread(fs.exists, path)

    async def extract(
        self, context: grpc.ServicerContext, task_template: TaskTemplate, inputs: Optional[LiteralMap] = None
    ) -> FileSensorConfig:
        return FileSensorConfig(**task_template.custom)


AgentRegistry.register(FileSensor())
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_

### CODE REVIEW

File: tests/flytekit/unit/core/test_map_task.py [hunk 1]
```diff
@@ -262,3 +262,23 @@ def test_map_task_resolver(serialization_settings):
     t = mtr.load_task(loader_args=args)
     assert t.python_interface.inputs == mt.python_interface.inputs
     assert t.python_interface.outputs == mt.python_interface.outputs
+
+
+@pytest.mark.parametrize(
```
Comment (kumare3) [inline tests/flytekit/unit/core/test_map_task.py:267]: too many files have changed. can you rebase?

File: flytekit/extend/backend/base_sensor.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from flytekit.extend.backend.base_agent import AgentBase
+from flytekit.models.literals import LiteralMap
+from flytekit.models.task import TaskTemplate
+
+
+class SensorBase(AgentBase):
+    def __init__(self, task_type: str):
+        super().__init__(task_type=task_type, asynchronous=True)
+
+    async def poke(self, path: str) -> bool:
```
Comment (kumare3) [inline flytekit/extend/backend/base_sensor.py:24]: this is not right? It should get some other information? This is ok for files only. What if I want to sense HivePartitions?

I think the api should be

```python
DerivedSensor(SensorBase[T]):
async def poke(self, v: T):
    ...
async def extract(self, context: grpc.ServicerContext,
        output_prefix: str,
        task_template: TaskTemplate,
        inputs: Optional[LiteralMap] = None) -> T:
    ....
```

Comment (kumare3) [reply flytekit/extend/backend/base_sensor.py:24]: Happy to discuss

Comment (pingsutw) [reply flytekit/extend/backend/base_sensor.py]: done

**Pull request closed without merging**